src/middlewared/middlewared/plugins/idmap.py

```diff
@@ -0,0 +1,1120 @@
+            "datastore.insert", self._config.datastore, data,
+            {
+                "prefix": self._config.datastore_prefix
+            },
+        )
+        return await self._get_instance(data['id'])
+
+    @accepts(
+        Int('id', required=True),
+        Patch(
+            "idmap_domaintobackend_create",
+            "idmap_domaintobackend_update",
+            ("attr", {"update": True})
+        )
+    )
+    async def do_update(self, id, data):
+        """
+        Update idmap to backend mapping by id.
+        """
+        old = await self._get_instance(id)
+        new = old.copy()
+        new.update(data)
+
+        new = await self.middleware.call('idmap.common_backend_compress', new)
+        verrors = ValidationErrors()
+        if new['domain'] in [dstype.DS_TYPE_LDAP.value, dstype.DS_TYPE_DEFAULT_DOMAIN.value]:
+            if new['idmap_backend'] not in ['ldap', 'tdb', 'tdb2']:
+                verrors.add(
+                    'domaintobackend_create.idmap_backend',
+                    f'idmap backend [{new["idmap_backend"]}] is not appropriate for the system domain type {dstype[new["domain"]]}'
+                )
+        if verrors:
+            raise verrors
+
+        await self.middleware.call(
+            'datastore.update',
+            self._config.datastore,
+            id,
+            new,
+            {'prefix': self._config.datastore_prefix}
+        )
+        updated_entry = await self._get_instance(id)
+        try:
+            await self.middleware.call('idmap.get_or_create_idmap_by_domain', updated_entry['domain']['domain_name'])
+        except Exception as e:
+            self.logger.debug('Failed to generate new idmap backend: %s', e)
+
+        return updated_entry
+
+    @accepts(Int('id'))
+    async def do_delete(self, id):
+        """
+        Delete idmap to backend mapping by id
+        """
+        entry = await self._get_instance(id)
+        if entry['domain']['id'] <= dstype['DS_TYPE_DEFAULT_DOMAIN'].value:
+            raise CallError(f'Deleting mapping for [{entry["domain"]["idmap_domain_name"]}] is not permitted.', errno.EPERM)
+        await self.middleware.call("datastore.delete", self._config.datastore, id)
+
+
+class IdmapADService(CRUDService):
+    class Config:
+        datastore = 'directoryservice.idmap_ad'
+        datastore_prefix = 'idmap_ad_'
+        namespace = 'idmap.ad'
+
+    @accepts(
+        Dict(
+            'idmap_ad_create',
+            Dict(
+                'domain',
+                Int('id'),
+                Str('idmap_domain_name'),
+                Str('idmap_domain_dns_domain_name'),
+            ),
+            Int('range_low', required=True, validators=[Range(min=1000, max=2147483647)]),
+            Int('range_high', required=True, validators=[Range(min=1000, max=2147483647)]),
+            Str('schema_mode'),
+            Bool('unix_primary_group'),
+            Bool('unix_nss_info'),
```

Comment:
We should have default values set for all `create` in the schema, otherwise it will use whatever is the database default and the API user would have no clue about that.